### PR TITLE
IDEM-67

### PR DESCRIPTION
**Pull request by brettvniekerk**
- [Link to ticket](https://github.com/dltxio/proxy.idem.com.au/issues/67)
- Added xero service and GPIB billing

### CODE REVIEW

**Comment by brettvniekerk** [comment]
For testing with the Xero demo company:
`XERO_TENANT_ID = "8a6ab782-27c1-4a45-9b8d-674a0b8ee603"`
`XERO_SALES_CODE = "000"`
`XERO_PRICE = 10`
`XERO_GPIB_ID = "eaac8dce-cdaf-4386-869e-146c2a2ecb86"`

File: src/services/XeroService.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public async sendInvoices(authToken: XeroTokenSet): Promise<string> {
+        try {
+            // set the auth token from the POST request body
+            this.client.setTokenSet(authToken);
+
+            // create initial xero constants that won't change
+            const contact: Contact = {
+                contactID: this.gpibId
```
Comment (xiangzhe0812) [inline src/services/XeroService.ts:52]: let's not hardcode gpibId here, contactID should pass in from controller

File: src/services/XeroService.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            };
+
+            // get requests from the request db
+            const totalRequests = await this.requestRepository.count({
+                where: {
+                    to: XeroClients.GPIB_NAME
```
Comment (xiangzhe0812) [inline src/services/XeroService.ts:58]: same here. pass in from controller

File: src/services/XeroService.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+            // construct invoices object that contains single invoice for vendor
+            const invoices: Invoices = {
+                invoices: [
+                    {
+                        type: Invoice.TypeEnum.ACCREC,
+                        contact,
+                        date: new Date().toLocaleDateString("en-US"),
+                        lineItems,
+                        reference: `${XeroClients.GPIB_NAME} Invoice`,
```
Comment (xiangzhe0812) [inline src/services/XeroService.ts:80]: same here "GPIB_NAME"

File: src/user/user.controller.ts [hunk 4]
```diff
@@ -244,4 +248,20 @@ export class UserController {
...
+    })
+    @ApiResponse({
+        status: HttpStatus.OK
+    })
+    @ApiResponse({
+        status: HttpStatus.BAD_REQUEST
+    })
+    @Post("send-invoices")
+    async sendInvoices(
+        @Body("authToken") authToken: XeroTokenSet
```
Comment (xiangzhe0812) [inline src/user/user.controller.ts:263]: add VendorEnum into request body

File: src/services/XeroService.ts (hunk 1 above)
Comment (brettvniekerk) [reply src/services/XeroService.ts:52]: 👍

File: src/services/XeroService.ts (hunk 2 above)
Comment (brettvniekerk) [reply src/services/XeroService.ts:58]: 👍

File: src/services/XeroService.ts (hunk 3 above)
Comment (brettvniekerk) [reply src/services/XeroService.ts:80]: 👍

File: src/user/user.controller.ts (hunk 4 above)
Comment (brettvniekerk) [reply src/user/user.controller.ts]: 👍

**Review by JasminFarhi — APPROVED** [review]

**Review by OptimusOpus — APPROVED** [review]
LGTM

File: src/services/XeroService.ts [hunk 5]
```diff
@@ -40,22 +41,35 @@ export class XeroService implements IXeroService {
...
         try {
             // set the auth token from the POST request body
             this.client.setTokenSet(authToken);
 
-            // create initial xero constants that won't change
+            // create xero contact
+            let contactID: string, contactName: VendorName;
+            switch (vendor) {
+                case VendorEnum.GPIB:
+                    contactID = this.gpibId;
```
[the remark this answers is not in the document]
Comment (brettvniekerk) [reply src/services/XeroService.ts:59]: 👍

**Pull request merged**